### DISCUSSION TITLE
V4.0 : Better for some

### DISCUSSION THREAD

**Discussion by easytarget** [easytarget/esp32-cam-webserver#236] [Announcements]
# The v4.0 release.
* Latest esp-arduino-core release (SDK v4.4-beta1)
* OTA enabled; configurable (including password) in `myconfig.h`
* MDNS support
* Camera bus xclock setting in config and UI, improves running on budget modules with bus issues, allows tuning the xclk frequency on the fly to tune for best speed/quality/etc
* SPIFFS is now read (and formatted if needed) before camera module initialisation; this helps with a SPIFFS startup errors
* 'kill stream' button in dump page, terminates the stream at the camera end in cases where the browser fails to drop the connection.
* UI tuneable frame rate limiter for smoother displays and or lower network use
* More info in `dump` page, including MCU temperature and xclk frequency
* UI tweaks, some notes and help for quality, xclk and other settings
* Improved display and logging of critical errors
* Support ArduCAM_ESP32S_UNO
* A whole bunch of spelling, comment and log tweaks

#### But..
* No more Face Recognition (actually; this is a feature, it allows for OTA updating)
* No more support for modules without PSRAM.

#### A big 'Thank You' to all the other people who contributed to this release with code, problem reports and discussion.

<hr /><em>This discussion was created from the release <a href='https://github.com/easytarget/esp32-cam-webserver/releases/tag/v4.0'>V4.0 : Better for some</a>.</em>

**Comment by tve** [comment]
Excellent! Thank you! Can't wait to try it out :-)

**Reply by voldemar2022** [reply]
Hello. When compiling I get an error:
```
Arduino: 1.8.19 (Windows Store 1.8.57.0) (Windows 10), Плата:"ESP32 Dev Module, Enabled, Minimal SPIFFS (1.9MB APP with OTA/190KB SPIFFS), 240MHz (WiFi/BT), QIO, 80MHz, 4MB (32Mb), 921600, None"

C:\Users\Acer\Documents\Arduino\esp32-cam-webserver\esp32-cam-webserver.ino: In function 'void StartCamera()':

esp32-cam-webserver:338:12: error: 'struct camera_config_t' has no member named 'grab_mode'

     config.grab_mode = CAMERA_GRAB_LATEST;

            ^
```
Etc..

**Reply by easytarget** [reply]
Please follow the readme to install the latest ESP32 Arduino core. You must have `2.0.2` for this release, earlier ESP core versions  will fail to build, giving this error.

**Comment by voldemar2022** [comment]
Thanks, updated the traffic and everything is fine. But I can't connect to the router. It is not clear how and where myconfig.sample.h wifi ssid and password are written:  

struct station stationList[] = {{"Keenetic-8129","mSj-X6A-UzK-VHP", true}};

/*
 * You can extend the stationList[] above with additional SSID+Password pairs

struct station stationList[] = {{"Keenetic-8129", "mSj-X6A-UzK-VHP", true},
                                {"ssid2", "pass2", true},
                                {"ssid3", "pass3", false}};


not connecting router! long pasword?

**Reply by easytarget** [reply]
Do not edit the myconfig._sample_.h file directly, copy it to `myconfig.h`as documented.

Your definition (on the first line) looks good, but editing the example in the comments wont have any effect..
Password length is not a factor

You need to look at the console output when it is trying to connect and see what the actual errors look like.

**Comment by ozett** [comment]
is there a binary of this version anywhere around?

**Reply by easytarget** [reply]
No. There is not.
.. Since 1) WiFi details are currently hard coded in when compiling.and 2) I don't want the hassle of maintaining it.

**Comment by Coyotus-Pos** [comment]
Hi. Is it possible to boot into AP mode and then from the web environment switch to station mode and connect to an existing network?

**Reply by sonofcy** [reply]
I think that describes the default behaviour. In other words do NOT modify the config, just compile and upload as it comes and that's what happens.

**Reply by easytarget** [reply]
No, The default is to start an AP and nothing else.

There is no connection manager built into the code, the only way to configure networks is to add and compile them in via `myconfig.h` when building the sketch.

Adding a network manager would be nice.. I need to get 4.1 with multi-stream support out and then I can look at the work that has been suggested in #212

**Reply by Coyotus-Pos** [reply]
Yes, that would be great, a wifi manager is all you need.

**Comment by terichardson** [comment]
Does anyone have any insight on using an OV5640 camera with this device? When I ran the sketch, the camera initiated but this message appeared:
07:35:10.185 -> WARNING: Camera module is unknown and not properly supported, will fallback to OV2640 operation
07:35:10.468 -> Preference file /esp32cam-preferences.json not found; using system defaults.
The sketch displayed an image, but the quality was very poor.
Thanks.

**Reply by tve** [reply]
You should start a new thread, unless you're specifically asking for the differences between V4 and prior versions... ;-)

I've tried OV5640 with prior versions and was underwhelmed. I got a ton of noise in the image. I don't know whether it's due to the sensors one can buy cheaply being crap or whether there are some magic settings the esp32 driver doesn't set correctly. I played around with it for a bit and could get to mediocre image quality at low frame rate. But none of this is really esp32-cam-webserver's fault.

**Comment by terichardson** [comment]
Thanks for the reply.  You just answered my question.  I just needed someone to confirm that version 4.0 is supposed to work with the OV5640. My experience thus far is that the cameras are poorly made.  No need for another thread. 
Tom

Thomas E. Richardson
256-565-9486

**Comment by ovywan** [comment]
I just discovered your work while looking for a stable ESP32-CAM code. It works flawlessly from the first try. Thank you.

I would be interested in viewing the stream from the ESP32-CAM on another ESP32, that has a 320×240 TFT display (ILI9341, available [here](https://www.wemos.cc/en/latest/d1_mini_shield/tft_2_4.html)).
Are you aware of any projects that do that?

**Reply by lohtse** [reply]
you may want to take a look here

https://matchboxscope.github.io/firmware/FLASH.html   and make sure you scroll through the page as there's links to various things of interest




ps the code works great with Unmodified cams so no need to worry about disassembling part.. I use this with my esp32 trail cams
